### DISCUSSION TITLE
Inheriting functions from other contracts

### DISCUSSION THREAD

**Discussion by ChefAharoni** [Cyfrin/foundry-full-course-cu#274] [Q&A] [answered]
Hi everyone,

I've also uploaded my question to [StackExchange](https://ethereum.stackexchange.com/questions/152199/getting-inner-erc20-functions-data-from-outer-erc20-deployments), but I'll try to shorten it here.

I have an ERC20 contract implementation that's inherited from OpenZeppelin's ERC20 token contract, and I want to use one of the implementations in another contract. 

What I initially thought of doing is re-writing the functions I want to use that are based in OpenZeppelin's ERC20 in **my** token contract, and then creating an instance of the token - but it just seems bad practice to me.

I tried writing an interface for my token contract, but I'm getting a weird error:
```
Derived contract must override function "_burn". 
Two or more base classes define function with same name and parameter types.solidity(6480)
```

This is the interface I wrote:
```solidity
interface IEcoCoin {
    function balanceOf(address _account) external view returns (uint256);
    function transferFrom(address _sender, address _recipient, uint256 _amount) external returns (bool);
    function _burn(address account, uint256 amount) external;
    function _mint(address _account, uint256 _amount) external;
}
```

This is the important part of **my** ERC20 token contract:
```solidity
import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
import {Muni, MuniData} from "./Municipality.sol";
import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";

contract EcoCoin is ERC20, MuniData, Ownable {
    using Muni for address;
    constructor() ERC20("EcoCoin", "ECC") {}

    function decimals() public pure override returns (uint8) {
        return 0; // No decimal places for the token; 2 tokens for a bottle, 10~ tokens equivalent to one usd.
    }

// This is an example of re-declaring the function from the ERC20 implementation;
// I've created this for every check I need, but it feels bad practice.

    function _burnTokens(
        address account,
        uint256 amount
    ) public muniOnly returns (bool) {
        _burn(account, amount); // Burn tokens
        return true;
    }
```

[This is my GitHub repo](https://github.com/ChefAharoni/EcoCoin) for reference. 

Thanks in advance for all the helpers.

**Comment by ghost** [comment]
EcoCoin is deriving `transferFrom`, `_mint`, `_burn` and `balanceOf` from ERC20 and IEcoCoin and the compiler see two sets of functions with the same names and types doesn't know what to do. I think the best thing to do is to remove `IEcoCoin` as a parent.

**Reply by ChefAharoni** [reply]
What do you mean remove as a parent?
Remove `IEcoCoin` from the `EcoCoin` contract declaration line?
Won't it disable the whole interface?

**Reply by ghost** [reply]
Sorry, I should've explained myself better.

Don't inherit `IEcoCoin` in `EcoCoin` because the functions in the interface are already inherited from the ERC20 standard. 

EcoCoin inherited two set of the same functions with the same types from two different "base contracts" and for this reason now you have to declare and override them in your derived contract (EcoCoin) which you're not doing, hence the error.

You'll have to do something like this:

```javascript
function transferFrom(address _sender, address _recipient, uint256 _amount) public override(IEcoCoin,  ERC20) returns (bool) {};
```

I read your post and checked your project (not super in depth lol) and since many functions were commented out maybe I go it all wrong but I don't fully understand why your already ERC20-compliant token will need to inherit an interface with the same functions.

**Reply by ChefAharoni** [reply]
First of all thank you for taking the time to observe and helping me, it is much appreciated. 

The functions I have commented out are the implementations of the function from OpenZeppelin's ERC20 contract, that I thought of using instead of the interface, such as this `burn` one:
```javascript
    /**
     * @notice  Function to burn tokens.
     * @dev     Uses the _burn function from the ERC20 contract, because it's internal.
     * @param   account  Account to burn tokens from.
     * @param   amount  Amount of tokens to burn.
     * @return  bool  True if the burn was successful.
     */
    // function _burnTokens(
    //     address account,
    //     uint256 amount
    // ) public muniOnly returns (bool) {
    //     _burn(account, amount); // Burn tokens
    //     return true;
    // }
```
<br>

If I understand correctly what you're saying, writing the function again in my `EcoCoin` contract, making it override the `ERC20` implementation, there's actually no use of an interface here, because I could just inherit my `EcoCoin` contract directly in other contracts, and use its functionality, right?

**Reply by ghost** [reply, marked as answer]
> If I understand correctly what you're saying, writing the function again in my EcoCoin contract, making it override the ERC20 implementation, there's actually no use of an interface here, because I could just inherit my EcoCoin contract directly in other contracts, and use its functionality, right?

Yes.

I see that in `depositor.sol` you imported `EcoCoin` to interact with the token. I what I would do is:

Import `IEcoCoin` but **not inherit** it in `Depositor`, deploy `EcoCoin` and pass the address of the token as a constructor argument in `Depositor` and create an instance of the token:

```javascript

IEcoCoin private immutable ecoCoin;

constructor(address _ecoCoin) {
     ecoCoin = IEcoCoin(_ecoCoin);
}

```

The above is my preference. 

Now, another thing to consider that I'm not sure about is whether it's more gas efficient to deploy the contract `Depositor` by importing the interface or the token contract.

**Reply by ChefAharoni** [reply]
That's actually very clever, thank you very much!

Seems it's working for now as it doesn't throw any errors, I'll write some tests in the future to ensure it's flawless.

Considering gas efficiency, I'm not _very_ concerned, as it's a college project with a deadline, and I'm trying to finish the basics before thinking of gas optimizing every small step; but I will notice it for future reference - thanks again!
